src/main/scala/io/iohk/ethereum/mining/Miner.scala

```diff
@@ -0,0 +1,194 @@
+    val epoch = Ethash.epoch(blockNumber.toLong)
+
+    val (dag, dagSize) = (currentEpoch, currentEpochDag, currentEpochDagSize) match {
+      case (Some(`epoch`), Some(dag), Some(dagSize)) => (dag, dagSize)
+      case _ =>
+        val seed = Ethash.seed(epoch)
+        val dagSize = Ethash.dagSize(epoch)
+        val dag =
+          if (!dagFile(seed).exists()) generateDagAndSaveToFile(epoch, dagSize, seed)
+          else loadDagFromFile(seed, dagSize)
+
+        currentEpoch = Some(epoch)
+        currentEpochDag = Some(dag)
+        currentEpochDagSize = Some(dagSize)
+      (dag, dagSize)
+    }
+
+    getBlockForMining(blockNumber) onComplete {
+      case Success(PendingBlock(block, _)) =>
+        val headerHash = crypto.kec256(BlockHeader.getEncodedWithoutNonce(block.header))
+        val mineResult = mine(headerHash, block.header.difficulty.toLong, dagSize, dag, miningConfig.mineRounds)
+        mineResult.foreach { case (pow, nonce) =>
+          syncController ! RegularSync.MinedBlock(block.copy(header = block.header.copy(nonce = nonce, mixHash = pow.mixHash)))
+        }
+        self ! ProcessMining
+
+      case Failure(ex) =>
+        log.error("Unable to get block for mining", ex)
+        context.system.scheduler.scheduleOnce(10.seconds, self, ProcessMining)
+    }
+  }
+
+  private def dagFile(seed: ByteString): File = {
+    new File(s"${miningConfig.ethashDir}/full-R${Ethash.Revision}-${Hex.toHexString(seed.take(8).toArray[Byte])}")
+  }
+
+  private def generateDagAndSaveToFile(epoch: Long, dagSize: Long, seed: ByteString): Array[Array[Int]] = {
+    // scalastyle:off magic.number
+
+    val file = dagFile(seed)
+    if (file.exists()) file.delete()
+    file.createNewFile()
+
+    val outputStream = new FileOutputStream(dagFile(seed).getAbsolutePath)
+    outputStream.write(Array(0xfe, 0xca, 0xdd, 0xba, 0xad, 0xde, 0xe1, 0xfe).map(_.toByte))
+
+    val cache = Ethash.makeCache(epoch)
+    val res = new Array[Array[Int]]((dagSize / Ethash.HASH_BYTES).toInt)
+
+    val max = (dagSize / Ethash.HASH_BYTES).toInt
+    (0 until max).foreach { i =>
+      val item = Ethash.calcDatasetItem(cache, i)
+      outputStream.write(Ethash.intsToBytes(item))
+      res(i) = item
+
+      if (i % 100000 == 0) log.info(s"Generating DAG ${((i / max.toDouble) * 100).toInt}%")
+    }
+
+    Try(outputStream.close())
+
+    res
+  }
+
+  private def loadDagFromFile(seed: ByteString, dagSize: Long): Array[Array[Int]] = {
+    val inputStream = new FileInputStream(dagFile(seed).getAbsolutePath)
+    inputStream.skip(8)
+    val buffer = new Array[Byte](64)
+    val res = new Array[Array[Int]]((dagSize / Ethash.HASH_BYTES).toInt)
+    var index = 0
+
+    while (inputStream.read(buffer) > 0) {
+      if (index % 100000 == 0) log.info(s"Loading DAG from file ${((index / res.length.toDouble) * 100).toInt}%")
+      res(index) = Ethash.bytesToInts(buffer)
+      index += 1
+    }
+
+    Try(inputStream.close())
+
+    res
```

Comment:
Maybe we could check whether the loaded file has the expected size? Maybe I'm missing something but it seems that if we stop the client while generating the DAG file it will load an only partially generated DAG.